### DISCUSSION TITLE
Prehashed signing and verification

### DISCUSSION THREAD

**Discussion by shriphani** [ZcashFoundation/ed25519-zebra#140] [Q&A] [answered]
Friends,

Thank you for your work on this library.

I can see some references to "prehash" in the repository but the comments seem to imply that it isn't what I am looking for. ed25519-dalek allows me to use the Sha512 digest of a potentially large message and sign / verify it using `signing_context.sign_digest` and `verifying_context.verify_digest` - is this functionality available in this crate? if not, is there any way to take advantage of the dalek implementation with this crate?

Best,
Shriphani

**Comment by conradoplg** [comment, marked as answer]
Hi,

We don't currently support contexts nor prehashed mode, but it's something that we could work on.

Just to be sure, is there a particular reason you need to use ed25519-zebra and not ed25519-dalek directly?

**Reply by shriphani** [reply]
Sorry, I took a while to reply.

In my system there is a potential backlog of signatures that builds up and I was wondering if I could store only the hash (or something similar) and not the full message which can get quite big. 

From what I understand (very cursorily), if I hash the message with something like SHA512 and then sign that, it is a little worse than signing the message itself because Ed25519 uses SHA512 internally for collision resistance and hashing the message would decrease the effectiveness of the collision resistance property.

If signing the hash is ok and not an anti-pattern, I guess I could just do that.

Thank you.

**Reply by conradoplg** [reply]
It's indeed not recommended to sign the hash without prehashed mode. What I was wondering is if you truly need the ZIP 215 variant of EdDSA which is provided by `ed25519-zebra`. If you don't need it, then you can just use `ed25519-dalek` where prehashed mode is available.

**Reply by shriphani** [reply]
I am using it in a consensus context where clients might bring their own libraries. I guess I can use the dalek crates and demand that all other clients should be able to accept and produce compatible signatures.
